src/builder.rs

```diff
@@ -556,6 +565,31 @@ impl<P: fmt::Debug, V> Bundle<InProgress<P, PartiallyAuthorized>, V> {
     }
 }
 
+impl<P: fmt::Debug, V> Bundle<InProgress<P, PartiallyAuthorized>, V> {
+    /// Appends an externally computed [`Signature`].
+    /// Signature index is inferred from signature randomizer `alpha`.
+    pub fn append_signature(
+        self,
+        signature: &redpallas::Signature<SpendAuth>,
+        alpha: &pallas::Scalar,
+        mut rng: impl RngCore,
+    ) -> Result<Self, Error> {
+        self.try_authorize(
+            &mut rng,
+            |_, partial, maybe| match maybe {
+                MaybeSigned::SigningMetadata(parts) if &parts.alpha == alpha => {
+                    let rk = parts.ak.randomize(&parts.alpha);
+                    rk.verify(&partial.sigs.sighash[..], signature)
+                        .map_err(Error::from)?;
+                    Ok(MaybeSigned::Signature(signature.clone()))
+                }
+                s => Ok(s),
+            },
+            |_, partial| Ok(partial),
+        )
+    }
```

Comment:
The `alpha` argument is unnecessary, because we can solely determine whether the signature applies to any output by trial-verifying in each output. If all signatures in the transaction are applied externally, this require `2n` trial signature verifications for `n` inputs (since we necessarily skip inputs that already have signatures). Given that we need `n` signature verifications anyway (since we don't want to append an invalid signature), I think this is an acceptable amount of work to simplify the API.

The `rng` argument is also unnecessary (as we aren't generating signatures); we can replace that with a context integer to determine error conditions.

```suggestion
    /// Appends an externally computed [`Signature`].
    ///
    /// The signature will be applied to the one input for which it is valid. An error
    /// will be returned if the signature is not valid for any inputs, or if it is valid
    /// for more than one input.
    pub fn append_signature(
        self,
        signature: &redpallas::Signature<SpendAuth>,
    ) -> Result<Self, Error> {
        let mut signature_valid_for = 0usize;
        self.try_authorize(
            &mut signature_valid_for,
            |valid_for, partial, maybe| match maybe {
                MaybeSigned::SigningMetadata(parts) => {
                    let rk = parts.ak.randomize(&parts.alpha);
                    if rk.verify(&partial.sigs.sighash[..], signature).is_ok() {
                        *valid_for += 1;
                        Ok(MaybeSigned::Signature(signature.clone()))
                    } else {
                        // Signature isn't for this input.
                        Ok(s)
                    }
                }
                s => Ok(s),
            },
            |_, partial| Ok(partial),
        ).expect("Closures cannot fail");
        match signature_valid_for {
            0 => Err(Error::InvalidExternalSignature),
            1 => Ok(()),
            _ => Err(Error::DuplicateSignature),
        }
    }
```

Reply:
You are right. That's a nice simplification.